zoekt-git-index: add support for delta shards

ggilmore opened the pull request:
Closes https://github.com/sourcegraph/zoekt/pull/259 

`zoekt-git-index` now supports a new CLI argument `-delta`. 

If this flag is provided, `zoekt-git-index` will first attempt to build the requested repository using the new delta shard strategy described in https://github.com/sourcegraph/sourcegraph/issues/29731. If this attempt fails , it will automatically fall back to try a "normal" (non-delta) build.

--- 

Note: This PR pulls in https://github.com/sourcegraph/zoekt/pull/259 as part of its changes, since looking up RepositoryMetadata() in this fashion is now a common operation.

ggilmore committed:
zoekt-git-index: add support for delta shards

zoekt-git-index now supports a new CLI argument "-delta". If this flag is provided, zoekt-git-index will first attempt to build the requested repository using the new delta shard strategy described in https://github.com/sourcegraph/sourcegraph/issues/29731. If this attempt fails , it will automatically fall back to try a "normal" (non-delta) build.

ggilmore committed:
builder e2e tests: explicitly sort branches while constructing input repository

(this was failing in CI, but not locally - I assume because Go's map iteration order is randomized)

ggilmore committed:
builder e2e_test.go: add back accidentally removed TestDetectLanguage

ggilmore committed:
index_test.go: add comments to delta tests

ggilmore committed:
index_test.go: fix grammar in one test step

ggilmore committed:
index_test.go: add more explanation in title for one test case

stefanhengl reviewed (commented):
First pass done. I will do another round tomorrow focussing more on prepareDeltaBuild.

keegancsmith commented on gitindex/index.go:
```diff
@@ -532,6 +539,229 @@ func newIgnoreMatcher(tree *object.Tree) (*ignore.Matcher, error) {
 	return ignore.ParseIgnoreFile(strings.NewReader(content))
 }
 
+type gitIndexConfig struct {
```
Yeah. Alternatively can't we setup git repos which emulate the behaviours we want to test for instead of mocking? That seems more robust. But unsure of the difficulty of that.

keegancsmith commented on gitindex/index.go:
```diff
@@ -403,36 +408,37 @@ func IndexGitRepo(opts Options) error {
+	var branchVersions map[string]map[string]plumbing.Hash
+
+	if opts.BuildOptions.IsDelta {
+		repos, branchMap, branchVersions, opts.BuildOptions.ChangedOrRemovedFiles, err = prepareDeltaBuildMetadata(opts, repo)
```
design: This doesn't feel like it belongs on BuildOptions. It would be like including documents in BuildOptions. Additionally the name seems very specific to git, but probably better than calling it Tombstone or something like that.

Consideration: Add a method to `Builder` which informs it of paths to Tombstone.

keegancsmith reviewed (changes requested):
Nice stuff. I did a thorough review, left some inline feedback :) Looking forward to landing this very soon!

stefanhengl commented on build/builder.go:
```diff
@@ -689,7 +719,14 @@ func (b *Builder) Finish() error {
+// of branch names and versions, IndexStateBranchSet if either slice specifies a different
+// set of branch names than the other, or IndexStateBranchVersion if both slices specify the same set of
+// branch names but have different accompanying versions.
+func CompareBranches(a, b []zoekt.RepositoryBranch) IndexState {
```
I am not sure I agree. I think gitindex is the most obvious call site and we can remove the call site from IndexState. Currently we don't switch based on the return state anyway.

ggilmore replied:
I'm a bit confused by the is comment chain. As keegan mentioned in https://github.com/sourcegraph/zoekt/pull/310#discussion_r835031449, there are many places that check for this information to pickup inconsistencies. This function is exported because these places exist both inside and outside the build package.

Are we okay in not enforcing invariants in multiple locations?

keegancsmith replied:
Discussion:

- Export `BranchesNamesEqual`
- Removing `IndexStateBranch*` states and tests
- keep checking invariant in crucial places that we do indeed have branch names equal.

stefanhengl commented on gitindex/index.go:
```diff
@@ -532,6 +539,229 @@ func newIgnoreMatcher(tree *object.Tree) (*ignore.Matcher, error) {
+
+				file := fileKey{Path: oldFile.Path(), ID: f.ID()}
+				repos[file] = hackSharedBlobLocation
+				branchMap[file] = append(branchMap[file], b)
```
I believe f.ID() will be different for all 3 versions. Since the key contains f.ID(), the branchMap for foo_2 should only contain main.

ggilmore replied:
I believe Stefan's explanation is correct - the ID should be different. 

I added a test to show that the existing behavior should be correct under your scenario: [`11a4e16` (#310)](https://github.com/sourcegraph/zoekt/pull/310/commits/11a4e16c5db4bf6cd9f32f70b64670c4bf77b6cd)

ggilmore committed:
extract prepareNormal/DeltaBuildMetadata declarations into separate types

ggilmore committed:
index.go: remove explicit gitconfig initialization from exported IndexGitRepo

ggilmore committed:
index.go: refactor to get rid of goto

ggilmore committed:
cmd/zoekt-git-index/main.go: log value of isDelta instead of conditional

ggilmore committed:
add test to see whether keegan's scenario in https://github.com/sourcegraph/zoekt/pull/310#discussion_r835171837 can occur

ggilmore committed:
remove commented out code

ggilmore committed:
prepareDeltaBuild: fail immediately if submodule indexing is turned on

ggilmore commented on build/builder.go:
```diff
@@ -222,6 +222,8 @@ type Builder struct {
 	// a sortable 20 chars long id.
 	id string
+
+	existingShards []string
```
Sorry, you're correct (but this check is still necessary). The failure mode isn't the shard writing, it's that `gitIndex` calls `b.Finish()` twice in the same run. Once at the end of the function, and another time in a defer statement: 

https://github.com/sourcegraph/zoekt/blob/86e6a54/gitindex/index.go#L466-L468 

The first time b.finish() is invoked, all the shards that it discovers through `oldShards := b.FindAllShards()` must be shards that existed prior to this latest job (since it hasn't renamed any of the new shards yet). So, it's safe to update the `FileTombstones` information on all of `oldShards`. 

However, the second time b.Finish() runs (as a result of the defer), without `existingShards` it can't disambiguate between the shards that it spat out during the previous b.Finish() invocation and shards that existed prior to this latest indexing job. In this case, it'll end up tombstoning files in the shards that it just created (which means that the latest version of those files are never discoverable). 

That's why we have to cache the list of previous existing shards in `existingShards` during builder's instantiation. This list is only written to once, and is never updated regardless of how many times you call `b.Finish()`. Only the shards in `existingShards`  should get their `filetombstone` metadata updated.

keegancsmith replied:
We discussed this over zoom, either:
- we make the idempotency of Finish much more explicit. IE a bool on builder that gates if it has been called. Then its an error to call Add if its true, etc.
- or documentation on existingShards use for idempotency.

ggilmore committed:
RepositoryMetadata: switch to using 'ok' boolean return value to communicate missing shards

ggilmore committed:
prepareDeltaBuild: switch to *Changes slice directly

keegancsmith commented on gitindex/index.go:
```diff
@@ -532,6 +539,229 @@ func newIgnoreMatcher(tree *object.Tree) (*ignore.Matcher, error) {
+	branchToCurrentTree := make(map[string]*object.Tree, len(options.Branches))
+
+	for _, b := range options.Branches {
+		commit, err := getCommit(repository, options.BranchPrefix, b)
+		if err != nil {
+			return nil, nil, nil, nil, fmt.Errorf("getting last current commit for branch %q: %w", b, err)
+		}
```
because any error we return means do normal build, we fallback to normal build and respect this option. Therefor we don't need to fail early since we do in fact support this option.

ggilmore committed:
index.go: ensure that reindexing same commits works with delta builds

ggilmore committed:
build/builder.go: get rid of CompareBranches in favor of BranchNamesEqual

- remove CompareBranches from IndexState()
- remove now unused IndexStateBranch[Version/Set]
- remove tests for IndexState (due to the above), and translate them into tests for BranchNamesEqual
- change former CompareBranches call-sites in builder.Finish() and gitindex.prepareDeltaBuild to use BranchNamesEqual

ggilmore committed:
builder.go: un-export changeOrRemovedFiles build option, use new MarkFileAsChangedOrRemoved() function instead

ggilmore commented on build/builder.go:
```diff
@@ -604,7 +634,7 @@ func (b *Builder) Finish() error {
 			}
 
-			if compareBranches(repository.Branches, b.opts.RepositoryDescription.Branches) == IndexStateBranchSet {
+			if CompareBranches(repository.Branches, b.opts.RepositoryDescription.Branches) == IndexStateBranchSet {
```
This is for enforcing the branch-name invariant (trying to catch bugs earlier)

ggilmore committed:
builder.go: fallback to normal build if a sourcegraph ignore file has been modified

I also discovered that go-git has really unintuitive handling of file names that come from diffs when the files are in a folder. After a lot of trial and error, I eventually stumbled-upon a "hack" to always get the full path relative to the repository's root: https://stackoverflow.com/a/58922399

I also added an extra test to make sure that this doesn't break in the future.

ggilmore commented on gitindex/index.go:
```diff
@@ -532,6 +539,229 @@ func newIgnoreMatcher(tree *object.Tree) (*ignore.Matcher, error) {
+	// loop over all branches, calculate the diff between our
+	// last indexed commit and the current commit, and add files mentioned in the diff
+	for _, branch := range existingRepository.Branches {
+		lastIndexedCommit, err := getCommit(repository, "", branch.Version)
```
Can we resolve this comment chain?

ggilmore committed:
modify folder test to have two documents: one in the root, one inside a folder

ggilmore committed:
RepositoryMetadata()-> FindRepositoryMetadata()

ggilmore committed:
builder.go: switch to using finishCalled boolean instead of existing shards slice

ggilmore committed:
builder_test.go: get rid of "HasSymbols" ignore field option

ggilmore commented:
PTAL - I believe that I responded to every comment.

ggilmore committed:
logs: "(delta build)" -> "delta build:"

keegancsmith reviewed (approved):
LGTM!

ggilmore committed:
builder.go: move MarkFileAsChangedOrRemoved to a builder method instead of an options method

ggilmore committed:
zoekt-sourcegraph-indexserver: add support for preparing delta builds behind a feature flag (#311)

ggilmore committed:
Merge remote-tracking branch 'origin/master' into zoekt-git-index-delta

ggilmore committed:
builder_test.go: disable ctags for TestFindRepositoryMetadata

ctags aren't relevant to this test, and they can cause the test to break depending on whether they're available.

ggilmore committed:
index_test.go: remove PR note from test
